src/blocks/battery/upower.rs

```diff
@@ -57,86 +56,176 @@ impl Device {
+                    .state()
                     .await
-                    .error("Failed to get time to empty")? as f64,
-            ),
-            _ => None,
-        };
+                    .error("Failed to get status")?
+                {
+                    1 => BatteryStatus::Charging,
+                    2 | 6 => BatteryStatus::Discharging,
+                    3 => BatteryStatus::Empty,
+                    4 => BatteryStatus::Full,
+                    5 => BatteryStatus::NotCharging,
+                    _ => BatteryStatus::Unknown,
+                };
+
+                let time_remaining = match status {
+                    BatteryStatus::Charging => Some(
+                        device_connection
+                            .device_proxy
+                            .time_to_full()
+                            .await
+                            .error("Failed to get time to full")? as f64,
+                    ),
+                    BatteryStatus::Discharging => Some(
+                        device_connection
+                            .device_proxy
+                            .time_to_empty()
+                            .await
+                            .error("Failed to get time to empty")? as f64,
+                    ),
+                    _ => None,
+                };
 
-        Ok(Some(BatteryInfo {
-            status,
-            capacity,
-            power: Some(power),
-            time_remaining,
-        }))
+                Ok(Some(BatteryInfo {
+                    status,
+                    capacity,
+                    power: Some(power),
+                    time_remaining,
+                }))
+            }
+        }
     }
 
     async fn wait_for_change(&mut self) -> Result<()> {
-        self.changes.next().await;
+        match &mut self.device_conn {
+            Some(device_connection) => {
+                select! {
+                    _ = device_connection.changes.next() => {},
+                    Some(msg)= self.device_removed_stream.next() => {
+                         let msg =  msg.unwrap();
+                        let body:  Structure = msg.body().unwrap();
+                        let fields = body.fields();
+                        if fields[0]==Value::ObjectPath(device_connection.device_path.clone()){
+                            self.device_conn = None;
+                        }
+
+                    },
+                }
+            }
+            None => {
+                if let Some(msg) = self.device_added_stream.next().await {
+                    let msg = msg.unwrap();
+                    let body: Structure = msg.body().unwrap();
+                    let fields = body.fields();
+                    if let Some(device_connection) =
+                        DeviceConnection::new(&self.dbus_conn, &self.device).await?
+                    {
+                        if fields[0] == Value::ObjectPath(device_connection.device_path.clone()) {
```

Comment:
Is this check actually necessary? The fact that `DeviceConnection::new` returned `Some` is enough I think.